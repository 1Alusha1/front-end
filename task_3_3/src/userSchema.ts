import { z } from "zod";

const specialCharacters = "!@#$%^&_";

const passwordSchema = z
  .string()
  .min(8, { message: "Password must be at least 8 characters long" })
  .refine((password) => /[A-Z]/.test(password), {
    message: "Password must contain at least one uppercase letter",
  })
  .refine((password) => /[a-z]/.test(password), {
    message: "Password must contain at least one lowercase letter",
  })
  .refine(
    (password) =>
      password.split("").some((char) => specialCharacters.includes(char)),
    { message: "Password must contain at least one special character" }
  );

export const userSchema = z.object({
  firstName: z.string().nonempty({ message: "First name is required" }),
  lastName: z.string().nonempty({ message: "Last name is required" }),
  email: z.string().email({ message: "Invalid email address" }),
  password: passwordSchema,
  confirmedPassword: z.string(),
  gender: z.string().nonempty({ message: "Gender is required" }),
  hobbies: z.array(z.enum(["Music", "Sports", "Travel", "Movies"])),
  sourceOfIncome: z
    .string()
    .nonempty({ message: "Source of income is required" }),
  income: z.number().min(1, { message: "Incom is required" }),
  upload: z.string().nonempty({ message: "Upload is required" }),
  age: z.number().min(18, { message: "You must be 18 years old" }),
  bio: z.string().nonempty({ message: "Bio is required" }),
});

export type User = z.infer<typeof userSchema>;
