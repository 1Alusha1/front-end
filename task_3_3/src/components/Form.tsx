import React, { useState } from "react";
import { userSchema, User } from "../userSchema";
import { z } from "zod";
import { isConfirmedPassword } from "../validation";
const Form: React.FC = () => {
  const [isSubmitted, setIsSubmitted] = useState(false);
  const [errors, setErrors] = useState<any | null>(null);
  const [user, setUser] = useState<User>({
    firstName: "",
    lastName: "",
    email: "",
    password: "",
    confirmedPassword: "",
    gender: "",
    hobbies: [],
    sourceOfIncome: "",
    income: 0,
    upload: "",
    age: 0,
    bio: "",
  });

  const setHobbies = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    if (
      ["Music", "Sports", "Travel", "Movies"].includes(
        value as (typeof user.hobbies)[number]
      )
    ) {
      setUser({
        ...user,
        hobbies: [...user.hobbies, value as (typeof user.hobbies)[number]],
      });
    }
  };

  const renderError = (name: string, errors: any[]) => {
    const error = errors?.find((error) => error.path === name);
    return error ? { error: true, message: error.message } : null;
  };

  const submitForm = (e: React.FormEvent) => {
    e.preventDefault();
    setIsSubmitted(true);

    try {
      const parsedUser = userSchema.parse(user);

      const confirmedPasswordValidation = isConfirmedPassword(
        parsedUser.password,
        parsedUser.confirmedPassword
      );

      if (!confirmedPasswordValidation.success) {
        throw confirmedPasswordValidation.error;
      }

      setErrors(null);
      console.log("Form submitted successfully!");
    } catch (error) {
      if (error instanceof z.ZodError) {
        setErrors(
          error.errors.map((error) => ({
            path: error.path[0],
            message: error.message,
          }))
        );
      } else if (error instanceof Error) {
        setErrors([{ path: "password", message: error.message }]);
      }
    }
  };
  return (
    <form onSubmit={submitForm}>
      <div className="full-name form-control">
        <div className="form-item">
          <label htmlFor="firstName">First name*</label>
          <input
            type="text"
            name="firstName"
            data-testid="FirstName"
            onChange={(e) => setUser({ ...user, firstName: e.target.value })}
            placeholder="Enter your first name"
            className={
              isSubmitted
                ? renderError("firstName", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          {isSubmitted && <p>{renderError("firstName", errors)?.message} </p>}
        </div>
        <div className="form-item">
          <label htmlFor="lastName">Last name*</label>
          <input
            type="text"
            name="lastName"
            onChange={(e) => setUser({ ...user, lastName: e.target.value })}
            placeholder="Enter your last name"
            className={
              isSubmitted
                ? renderError("lastName", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          {isSubmitted && <p>{renderError("lastName", errors)?.message} </p>}
        </div>
      </div>

      <div className="email form-control">
        <div className="form-item">
          <label htmlFor="email">Email*</label>

          <input
            type="email"
            onChange={(e) => setUser({ ...user, email: e.target.value })}
            name="email"
            placeholder="Enter your email"
            className={
              isSubmitted
                ? renderError("email", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          {isSubmitted && <p>{renderError("email", errors)?.message} </p>}
        </div>
      </div>

      <div className="password form-control">
        <div className="form-item">
          <label htmlFor="password">Password*</label>
          <input
            type="password"
            name="password"
            onChange={(e) => setUser({ ...user, password: e.target.value })}
            placeholder="Enter your password"
            className={
              isSubmitted
                ? renderError("password", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          {isSubmitted && <p>{renderError("password", errors)?.message} </p>}
        </div>
        <div className="form-item">
          <label htmlFor="confirmedPassword">Confirm Password*</label>
          <input
            type="password"
            onChange={(e) =>
              setUser({ ...user, confirmedPassword: e.target.value })
            }
            name="confirmedPassword"
            placeholder="Please repeat your confirm password"
            className={
              isSubmitted
                ? renderError("confirmedPassword", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          {isSubmitted && (
            <p>{renderError("confirmedPassword", errors)?.message} </p>
          )}
        </div>
      </div>
      <div className="group">
        <div className="gender form-control">
          <fieldset>
            <legend>Gender:</legend>
            <div className="gender-wrap">
              <div className="radio">
                <input
                  type="radio"
                  onChange={(e) => setUser({ ...user, gender: e.target.value })}
                  name="gender"
                  value="Male"
                  id="Male"
                />
                <label htmlFor="Male">Male</label>
              </div>
              <div className="radio">
                <input
                  type="radio"
                  onChange={(e) => setUser({ ...user, gender: e.target.value })}
                  name="gender"
                  value="Female"
                  id="Female"
                />
                <label htmlFor="Female">Female</label>
              </div>
            </div>
          </fieldset>
          {isSubmitted && (
            <p className="error">{renderError("gender", errors)?.message} </p>
          )}
        </div>
        <div className="hobbies form-control">
          <fieldset>
            <legend>Hobbies:</legend>
            <div className="hobbies-wrap">
              <div className="radio">
                <input
                  type="checkbox"
                  name="hobbies"
                  onChange={setHobbies}
                  value="Music"
                  id="Music"
                  className={
                    isSubmitted
                      ? renderError("hobbies", errors)?.error
                        ? "error"
                        : "success"
                      : ""
                  }
                />
                <label htmlFor="Music">Music</label>
              </div>
              <div className="radio">
                <input
                  type="checkbox"
                  name="hobbies"
                  onChange={setHobbies}
                  value="Sports"
                  id="Sports"
                  className={
                    isSubmitted
                      ? renderError("hobbies", errors)?.error
                        ? "error"
                        : "success"
                      : ""
                  }
                />
                <label htmlFor="Sports">Sports</label>
              </div>
              <div className="radio">
                <input
                  type="checkbox"
                  name="hobbies"
                  onChange={setHobbies}
                  value="Travel"
                  id="Travel"
                  className={
                    isSubmitted
                      ? renderError("hobbies", errors)?.error
                        ? "error"
                        : "success"
                      : ""
                  }
                />
                <label htmlFor="Travel">Travel</label>
              </div>
              <div className="radio">
                <input
                  type="checkbox"
                  name="hobbies"
                  onChange={setHobbies}
                  value="Movies"
                  id="Movies"
                  className={
                    isSubmitted
                      ? renderError("hobbies", errors)?.error
                        ? "error"
                        : "success"
                      : ""
                  }
                />
                <label htmlFor="Movies">Movies</label>
              </div>
            </div>
          </fieldset>
          {isSubmitted && (
            <p className="error">{renderError("hobbies", errors)?.message} </p>
          )}
        </div>
      </div>

      <div className="income form-control">
        <div className="form-item">
          <label htmlFor="sourceOfIncome">Source of Income</label>
          <select
            onChange={(e) =>
              setUser({ ...user, sourceOfIncome: e.target.value })
            }
            name="sourceOfIncome"
            id="sourceOfIncome"
            data-testid="sourceOfIncome"
            className={
              isSubmitted
                ? renderError("sourceOfIncome", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          >
            <option value="Employed">Employed</option>
            <option value="Employed">Employed</option>
            <option value="Employed">Employed</option>
            <option value="Employed">Employed</option>
          </select>
          {isSubmitted && (
            <p className="error">
              {renderError("sourceOfIncome", errors)?.message}{" "}
            </p>
          )}
        </div>
        <div className="form-item income-l">
          <label htmlFor="income">Income</label>
          <input
            type="range"
            min="0"
            max="100"
            name="income"
            data-testid="income"
            onChange={(e) => setUser({ ...user, income: +e.target.value })}
            id="income"
            className={
              isSubmitted
                ? renderError("income", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          <button className="btn">20K</button>
        </div>
        {isSubmitted && (
          <p className="error income-l-e">
            {renderError("income", errors)?.message}{" "}
          </p>
        )}
      </div>

      <div className="upload form-control ">
        <div className="form-item">
          <label htmlFor="upload">Upload Profile Picture</label>
          <div className="upload-g">
            <input
              type="file"
              onChange={(e) =>
                setUser({ ...user, upload: JSON.stringify(e.target.files) })
              }
              name="upload"
              id="upload"
              data-testid="upload"
              className={
                isSubmitted
                  ? renderError("upload", errors)?.error
                    ? "error"
                    : "success"
                  : ""
              }
            />
            <button className="btn btn-dark">Choose file</button>
            <span>No file chosen</span>
          </div>
          {isSubmitted && (
            <p className="error">{renderError("upload", errors)?.message} </p>
          )}
        </div>
        <div className="form-item ">
          <label htmlFor="age">Age</label>
          <input
            type="number"
            onChange={(e) => setUser({ ...user, age: Number(e.target.value) })}
            name="age"
            id="age"
            data-testid="age"
            className={
              isSubmitted
                ? renderError("age", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          />
          {isSubmitted && (
            <p className="error">{renderError("age", errors)?.message} </p>
          )}
        </div>
      </div>

      <div className="about form-control">
        <div className="form-item">
          <label htmlFor="bio">Bio</label>
          <textarea
            name="bio"
            onChange={(e) => setUser({ ...user, bio: e.target.value })}
            placeholder="Tell about yourself"
            cols={30}
            rows={10}
            className={
              isSubmitted
                ? renderError("bio", errors)?.error
                  ? "error"
                  : "success"
                : ""
            }
          ></textarea>
          {isSubmitted && (
            <p className="error">{renderError("bio", errors)?.message} </p>
          )}
        </div>
      </div>
      <div className="submit form-control">
        <div className="form-item">
          <input type="submit" className="btn" value="Create" data-testid="create" />
        </div>
      </div>
    </form>
  );
};

export default Form;
