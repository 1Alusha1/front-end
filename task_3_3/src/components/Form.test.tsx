import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
// import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Form from "./Form";
import { userSchema } from "../userSchema";
import { z } from "zod";

jest.mock("../userSchema", () => {
  return {
    userSchema: {
      parse: jest.fn(),
    },
  };
});

jest.mock("../validation", () => ({
  isConfirmedPassword: jest.fn(),
}));

describe("Form Component", () => {

  test("submits the form", async () => {
    (userSchema.parse as jest.Mock).mockImplementation(() => ({
      firstName: "John",
      lastName: "Doe",
      email: "john.doe@example.com",
      password: "Password1!",
      confirmedPassword: "Password1!",
      gender: "Male",
      hobbies: ["Music", "Movies"],
      sourceOfIncome: "Employed",
      income: 100,
      upload: "profile.jpg",
      age: 30,
      bio: "Hello, I am John!",
    }));

    const { isConfirmedPassword } = require("../validation");
    (isConfirmedPassword as jest.Mock).mockReturnValue({ success: true });

    render(<Form />);

    fireEvent.change(screen.getByPlaceholderText("Enter your first name"), {
      target: { value: "John" },
    });
    fireEvent.change(screen.getByPlaceholderText("Enter your last name"), {
      target: { value: "Doe" },
    });
    fireEvent.change(screen.getByPlaceholderText("Enter your email"), {
      target: { value: "john.doe@example.com" },
    });
    fireEvent.change(screen.getByPlaceholderText("Enter your password"), {
      target: { value: "Password1!" },
    });
    fireEvent.change(
      screen.getByPlaceholderText("Please repeat your confirm password"),
      {
        target: { value: "Password1!" },
      }
    );
    fireEvent.click(screen.getByLabelText("Male"));
    fireEvent.click(screen.getByLabelText("Music"));
    fireEvent.click(screen.getByLabelText("Movies"));

    fireEvent.change(screen.getByTestId("sourceOfIncome"), {
      target: { value: "Employed" },
    });
    fireEvent.change(screen.getByTestId("income"), {
      target: { value: 100 },
    });
    fireEvent.change(screen.getByTestId("upload"), {
      target: { files: ["profile.jpg"] },
    });
    fireEvent.change(screen.getByTestId("age"), {
      target: { value: "30" },
    });
    fireEvent.change(screen.getByPlaceholderText("Tell about yourself"), {
      target: { value: "Hello, I am John!" },
    });

    fireEvent.click(screen.getByTestId("create"));

    expect(userSchema.parse).toHaveBeenCalledWith({
      firstName: "John",
      lastName: "Doe",
      email: "john.doe@example.com",
      password: "Password1!",
      confirmedPassword: "Password1!",
      gender: "Male",
      hobbies: ["Music", "Movies"],
      sourceOfIncome: "Employed",
      income: 100,
      upload: '["profile.jpg"]',
      age: 30,
      bio: "Hello, I am John!",
    });

    expect(isConfirmedPassword).toHaveBeenCalledWith(
      "Password1!",
      "Password1!"
    );
  });

  test("shows validation errors on invalid input", async () => {
    (userSchema.parse as jest.Mock).mockImplementation(() => {
      throw new z.ZodError([
        {
          path: ["firstName"],
          message: "First name is required",
          code: "custom",
        },
        { path: ["email"], message: "Invalid email address", code: "custom" },
      ]);
    });

    render(<Form />);

    fireEvent.click(screen.getByText(/Create/i));

    expect(
      await screen.findByText("First name is required")
    ).toBeInTheDocument();
    expect(
      await screen.findByText("Invalid email address")
    ).toBeInTheDocument();
  });
});
