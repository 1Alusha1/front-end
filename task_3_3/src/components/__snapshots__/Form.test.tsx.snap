// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test app loads and displays greeting 1`] = `
<input
  class=""
  name="email"
  placeholder="Enter your email"
  type="email"
/>
`;

exports[`loads and displays greeting 1`] = `Promise {}`;
